Saule/Http/PreprocessingDelegatingHandler.cs

```diff
@@ -43,11 +44,15 @@ public PreprocessingDelegatingHandler(JsonApiConfiguration config)
 
             ApiResource resource = null;
             bool isHttpError = content is HttpError || content is IEnumerable<HttpError>;
-            if (request.Properties.ContainsKey(Constants.PropertyNames.ResourceDescriptor))
+            IApiResourceProvider resourceProvider = null;
+
+            if (!isHttpError)
             {
-                resource = (ApiResource)request.Properties[Constants.PropertyNames.ResourceDescriptor];
+                resourceProvider = config.ApiResourceProviderFactory.Create(request);
+                resource = resourceProvider?.Resolve(content);
```

Comment:
I think it's probably fair to throw if the factory returns null? I don't know if there's a scenario where that makes sense, and we can create a nice error message at this spot that allows people to figure out what went wrong.

Reply:
i had it initially to return null and fail in that case, but i found that there are unit tests like this - https://github.com/joukevandermaas/saule/blob/6d79f830f7442631e373ee581d702b01131dfb94/Tests/Controllers/CompaniesController.cs#L25-L31 . we don't have `ReturnsResource` here so no active `ApiResource` and code works fine as we don't have any `httpContent` as result and it's just `HttpStatusCode.NoContent`. 

from my understanding it makes sense to fail if it's return null as `ApiResource` always should be available, but then it would break that case when Endpoint doesn't return any object and doesn't need `ApiResource` so i updated it to allow null `ApiResource`

Reply:
Alright, I think that's a good reason. I don't want to break people's existing code that may depend on this.

Do you mind adding a comment to explain this?

Reply:
yep, will do

Reply:
so i checked it more and added a check for non null ApiResourceProvider. ApiResourceProvider itself might return a null ApiResource but it can't be null itself. Added a comment to the interface that mentions it. Also added an error if it's a null. And also added unit test to cover that scenario